### ISSUE TITLE
GetWorkflowStateAsync throws an exception when instanceid is not found

### ISSUE THREAD

**Issue by bgelens** [dapr/dotnet-sdk#1195]
## Expected Behavior

When calling `GetWorkflowStateAsync` and the instanceid is not found, I expect it to return a `WorkflowState` with the property `Exists` being `false`

Not sure yet about my approach here yet but I'm trying to schedule a workflow during startup but only if it is not yet scheduled and / or its inputs have changed. (Basically replacing a backgroundservice by a workflow that calls ContinueAsNew

## Actual Behavior

Grpc.Core.RpcException: Status(StatusCode="Unknown", Detail="no such instance exists")

## Steps to Reproduce the Problem

```c#
var myWorkflow = await daprWorkflowClient.GetWorkflowStateAsync("myNonExistingWorkflow");
myWorkflow.Exists == false;
```

## Release Note

RELEASE NOTE:

**Comment by halspang** [comment]
@cgillum @RyanLettieri - Is an exception on a non-existent workflow the intended behavior or should we change it to be this empty/boolean response for if it existed or not?

**Comment by cgillum** [comment]
@halspang this is not the intended behavior for the SDKs. It's an unintended side-effect of how the APIs in the runtime work.

**Comment by halspang** [comment]
@cgillum - Is this fully a runtime issue then? Or should we just be intercepting the specific error that the runtime returns instead of throwing an exception in the SDK?

**Comment by cgillum** [comment]
@halspang possibly, but it's something we may need to discuss. Part of the feedback we're getting is that these show up as errors in the Dapr runtime logs, which is super noisy, so we likely will want to make this change in the runtime. Of course, we'll also need to make sure that the client handles this case properly after we make any runtime changes.

**Comment by richteaclue** [comment]
@halspang @cgillum there is a secondary issue here, beyond the question about whether to throw an exception vs an empty/boolean response.

In the [deprecated](https://github.com/dapr/dotnet-sdk/issues/1205) version of the API (based on `DaprClient` not `DaprWorkflowClient`) a non-existent `instanceId` leads to a `DaprException` that wraps an `RpcException` with `StatusCode == NotFound`. In the `DaprWorkflowClient` version, it leads to a raw `RpcException`, but crucially this exception has `StatusCode == Unknown` (as reported in the OP), which makes it difficult to handle programmatically. The SDK consumer has a choice between either examining the `Detail` text (ugh!) or using the deprecated `DaprClient` API (also ugh!).

Would it be possible for now to modify the new SDK to ensure that the `RpcException` has `StatusCode == NotFound`, which can at least be handled without reading information that is intended for humans?

I'm happy to look at creating a fix PR if someone could give me a pointer as to which repo to fix - it looks like `DaprClient` and `DaprWorkflowClient` make different RPC calls, so I'd guess that the fix is in the sidecar repo?

**Comment by cgillum** [comment]
> Would it be possible for now to modify the new SDK to ensure that the RpcException has StatusCode == NotFound, which can at least be handled without reading information that is intended for humans?

I think it would make sense to fix this in the Dapr runtime, which implements the gRPC API. @ItalyPaleAle do you know whether the universal API allows us to specify a `StatusCode` value of `NotFound`? This would be for the workflow management APIs that @RyanLettieri created.

**Comment by ItalyPaleAle** [comment]
Yes if you respond with a messages.APIError the built-in status code is honored. 

https://github.com/dapr/dapr/blob/master/pkg/http/responses.go#L142-L146

https://github.com/dapr/dapr/blob/master/pkg/messages/predefined.go

**Comment by cgillum** [comment]
Marking this as P0 since the current behavior is problematic and we need to lock down the API surface area.

One note about this, however, is that we're planning to remove usage of these APIs from the generic Dapr SDKs and instead rely on the Dapr _Workflow_ SDKs for managing workflows. The Dapr Workflow SDKs use a different set of APIs which aren't affected by this issue. However, I think we still want to prioritize fixing this for HTTP clients that want to check the status of workflows. The important thing in those cases is that we return HTTP 404 instead of HTTP 500. Hopefully this is a very small and simple change.
